### PR TITLE
Support filter for List

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Close https://github.com/apache/datafusion/issues/9857

### CODE REVIEW

**Comment by jayzhan211** [comment]
It seems we need to rewrite `array || element` to `array_append` first

File: datafusion/physical-expr-common/src/datum.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    };
+
+    // fast path, if compare with one null and operator is not 'distinct', then we can return null array directly
+    if !matches!(op, Operator::IsDistinctFrom | Operator::IsNotDistinctFrom)
+        && (is_l_scalar && l.null_count() == 1 || is_r_scalar && r.null_count() == 1)
+    {
+        return Ok(BooleanArray::new_null(len));
+    }
+
+    let cmp = make_comparator(l, r, SortOptions::default())?;
```
Comment (alamb) [inline datafusion/physical-expr-common/src/datum.rs:116]: is it worth a comment here explaining the choice of NULL comparison / sorting?

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211  -- this looks very good to me 👨‍🍳 👌 

Really nice

**Comment by alamb** [comment]
🚀

**Pull request merged**